pkg/controller/core.oam.dev/v1alpha1/envbinding/cluster_gateway_engine.go

```diff
@@ -0,0 +1,112 @@
+/*
+ Copyright 2021. The KubeVela Authors.
+
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+     http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+*/
+
+package envbinding
+
+import (
+	"context"
+
+	"github.com/pkg/errors"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/oam-dev/kubevela/apis/core.oam.dev/v1alpha1"
+	"github.com/oam-dev/kubevela/apis/core.oam.dev/v1beta1"
+	"github.com/oam-dev/kubevela/pkg/appfile"
+	"github.com/oam-dev/kubevela/pkg/multicluster"
+)
+
+// ClusterGatewayEngine construct the multicluster engine of using cluster-gateway
+type ClusterGatewayEngine struct {
```

Comment:
It would be better to provide an interface of ClusterGatewayEngine.
In this way it will have clear API cuts and would easier to write unit test.

Reply:
https://github.com/Somefive/kubevela/blob/5d989fa5f9e0d083e7b244c9b6ebad34573d8c38/pkg/controller/core.oam.dev/v1alpha1/envbinding/engine.go#L42

sure, we have already provide an interface